File: tests/scalers/cassandra.test.ts
import test from 'ava'
import * as fs from 'fs'
import * as sh from 'shelljs'
import * as tmp from 'tmp'
import { createNamespace } from './helpers'

const cassandraNamespace = 'cassandra-test'
const cassandraKeyspace = 'test_keyspace'
const cassandraTableName =  'test_table'
const cassandraUsername = 'cassandra'
const cassandraPassword = 'cassandra'
const nginxDeploymentName = 'nginx-deployment'

test.before(t => {
    // install cassandra
    console.log("Install cassandra")
    createNamespace(cassandraNamespace)
    const cassandraTmpFile = tmp.fileSync()
    fs.writeFileSync(cassandraTmpFile.name, cassandraDeployYaml)

    t.is(0, sh.exec(`kubectl apply --namespace ${cassandraNamespace} -f ${cassandraTmpFile.name}`).code, 'creating a Cassandra deployment should work.')

    // wait for cassandra to load
    console.log("wait for cassandra to load")
    let cassandraReadyReplicaCount = '0'
     for (let i = 0; i < 30; i++) {
        cassandraReadyReplicaCount = sh.exec(`kubectl get deploy/cassandra -n ${cassandraNamespace} -o jsonpath='{.status.readyReplicas}'`).stdout
        if (cassandraReadyReplicaCount != '1') {
            sh.exec('sleep 10s')
        }
    }
    t.is('1', cassandraReadyReplicaCount, 'Cassandra is not in a ready state')

    console.log("cassandraReadyCheck")
    let cassandraReady
    const cassandraPod = sh.exec(`kubectl get pods --selector=app=cassandra-app -n ${cassandraNamespace} -o jsonpath='{.items[0].metadata.name}'`).stdout
    for (let i = 0; i < 30; i++) {
      cassandraReady = sh.exec(`kubectl exec -n ${cassandraNamespace} ${cassandraPod} -- nodetool status | grep -w -o UN`)
      if (cassandraReady != "UN\n") {
        sh.exec('sleep 10s')
      }
      else {
        break
      }
    }

    // create cassandra-client
    console.log("create cassandra-client")
    const cassandraClientTmpFile = tmp.fileSync()
    fs.writeFileSync(cassandraClientTmpFile.name, cassandraClientDeployYaml)

    t.is(0, sh.exec(`kubectl apply --namespace ${cassandraNamespace} -f ${cassandraClientTmpFile.name}`).code, 'creating a Cassandra client deployment should work.')

    // wait for cassandra-client to load
    console.log("wait for cassandra-client to load")
    let cassandraClientReadyReplicaCount = '0'
    for (let i = 0; i < 30; i++) {
      cassandraClientReadyReplicaCount = sh.exec(`kubectl get deploy/cassandra-client -n ${cassandraNamespace} -o jsonpath='{.status.readyReplicas}'`).stdout
      if (cassandraClientReadyReplicaCount != '1') {
        sh.exec('sleep 10s')
      }
    }
    t.is('1', cassandraClientReadyReplicaCount, 'Cassandra client is not in a ready state')

    console.log("cassandraClientReadyCheck")
    let cassandraClientReady
    const cassandraClientPod = sh.exec(`kubectl get pods --selector=app=cassandra-client -n ${cassandraNamespace} -o jsonpath='{.items[0].metadata.name}'`).stdout
    for (let i = 0; i < 30; i++) {
      cassandraClientReady = sh.exec(`kubectl exec -n ${cassandraNamespace} ${cassandraClientPod} -- nodetool status | grep -w -o UN`)
      if (cassandraClientReady != "UN\n") {
        sh.exec('sleep 10s')
      }
      else {
        break
      }
    }

    // create table
    console.log("create table")
    const createKeyspace = `CREATE KEYSPACE IF NOT EXISTS ${cassandraKeyspace} WITH REPLICATION = {'class' : 'NetworkTopologyStrategy', 'datacenter1' : '1'};`
    const createTableCQL = `CREATE TABLE IF NOT EXISTS ${cassandraKeyspace}.${cassandraTableName} (name text, surname text, age int, PRIMARY KEY (name, surname));`
    t.not(cassandraClientPod, '')
    sh.exec(`kubectl exec ${cassandraClientPod} -n ${cassandraNamespace} -- bash cqlsh -u ${cassandraUsername} -p ${cassandraPassword} cassandra.${cassandraNamespace} --execute="${createKeyspace}"`)
    sh.exec(`kubectl exec ${cassandraClientPod} -n ${cassandraNamespace} -- bash cqlsh -u ${cassandraUsername} -p ${cassandraPassword} cassandra.${cassandraNamespace} --execute="${createTableCQL}"`)

    // deploy nginx, scaledobject etc.
    console.log("deploy nginx, scaledobject etc.")
    const nginxTmpFile = tmp.fileSync()
    fs.writeFileSync(nginxTmpFile.name, nginxDeployYaml)

    t.is(0, sh.exec(`kubectl apply --namespace ${cassandraNamespace} -f ${nginxTmpFile.name}`).code, 'creating nginx deployment should work.')
    // wait for nginx to load
    console.log("wait for nginx to load")
    let nginxReadyReplicaCount = '0'
    for (let i = 0; i < 30; i++) {
      nginxReadyReplicaCount = sh.exec(`kubectl get deploy/${nginxDeploymentName} -n ${cassandraNamespace} -o jsonpath='{.status.readyReplicas}'`).stdout
      if (nginxReadyReplicaCount != '') {
        sh.exec('sleep 10s')
      }
    }
    t.is('', nginxReadyReplicaCount, 'creating an Nginx deployment should work')

})

test.serial('Should start off deployment with 0 replicas', t => {

    const replicaCount = sh.exec(`kubectl get deploy/${nginxDeploymentName} --namespace ${cassandraNamespace} -o jsonpath="{.spec.replicas}"`).stdout
    t.is(replicaCount, '0', 'Replica count should start out as 0')

})

test.serial(`Replicas should scale to 2 (the max) then back to 0`, t => {
    // insert data to cassandra
    console.log("insert data to cassandra")
    const insertData = `BEGIN BATCH
    INSERT INTO ${cassandraKeyspace}.${cassandraTableName} (name, surname, age) VALUES ('Mary', 'Paul', 30);
    INSERT INTO ${cassandraKeyspace}.${cassandraTableName} (name, surname, age) VALUES ('James', 'Miller', 25);
    INSERT INTO ${cassandraKeyspace}.${cassandraTableName} (name, surname, age) VALUES ('Lisa', 'Wilson', 29);
    INSERT INTO ${cassandraKeyspace}.${cassandraTableName} (name, surname, age) VALUES ('Bob', 'Taylor', 33);
    INSERT INTO ${cassandraKeyspace}.${cassandraTableName} (name, surname, age) VALUES ('Carol', 'Moore', 31);
    INSERT INTO ${cassandraKeyspace}.${cassandraTableName} (name, surname, age) VALUES ('Richard', 'Brown', 23);
    APPLY BATCH;`

    const cassandraClientPod = sh.exec(`kubectl get pods --selector=app=cassandra-client -n ${cassandraNamespace} -o jsonpath='{.items[0].metadata.name}'`).stdout
    t.not(cassandraClientPod, '')

    t.is(
        0,
        sh.exec(`kubectl exec ${cassandraClientPod} -n ${cassandraNamespace} -- bash cqlsh -u ${cassandraUsername} -p ${cassandraPassword} cassandra.${cassandraNamespace} --execute="${insertData}"`).code,
        'insert 6 cassandra record'
    )

    let replicaCount = '0'
    const maxReplicaCount = '2'

    for (let i = 0; i < 30 && replicaCount !== maxReplicaCount; i++) {
      replicaCount = sh.exec(
        `kubectl get deploy/${nginxDeploymentName} --namespace ${cassandraNamespace} -o jsonpath="{.spec.replicas}"`).stdout
        if (replicaCount !== maxReplicaCount) {
          sh.exec('sleep 10s')
        }
    }

    t.is(maxReplicaCount, replicaCount, `Replica count should be ${maxReplicaCount} after 300 seconds`)
    sh.exec('sleep 30s')

    // delete all data from cassandra
    console.log("delete all data from cassandra")
    const truncateData = `TRUNCATE ${cassandraKeyspace}.${cassandraTableName};`

    t.is(
        0,
        sh.exec(`kubectl exec ${cassandraClientPod} -n ${cassandraNamespace} -- bash cqlsh -u ${cassandraUsername} -p ${cassandraPassword} cassandra.${cassandraNamespace} --execute="${truncateData}"`).code,
        'delete all rows'
    )

    for (let i = 0; i < 30 && replicaCount !== '0'; i++) {
      replicaCount = sh.exec(
        `kubectl get deploy/${nginxDeploymentName} --namespace ${cassandraNamespace} -o jsonpath="{.spec.replicas}"`).stdout
      if (replicaCount !== '0') {
        sh.exec('sleep 10s')
      }
    }

     t.is('0', replicaCount, 'Replica count should be 0 after 5 minutes')

})

test.after.always((t) => {
     t.is(0, sh.exec(`kubectl delete namespace ${cassandraNamespace}`).code, 'Should delete Cassandra namespace')

})

const cassandraDeployYaml = `
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app: cassandra-app
  name: cassandra
spec:
  replicas: 1
  selector:
    matchLabels:
      app: cassandra-app
  template:
    metadata:
      labels:
        app: cassandra-app
    spec:
      containers:
      - image: bitnami/cassandra:4.0.1
        imagePullPolicy: IfNotPresent
        name: cassandra
        ports:
        - containerPort: 9042
---
apiVersion: v1
kind: Service
metadata:
    name: cassandra
spec:
    ports:
      - name: cql
        port: 9042
        protocol: TCP
        targetPort: 9042
    selector:
        app: cassandra-app
`

const cassandraClientDeployYaml = `
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app: cassandra-client
  name: cassandra-client
spec:
  replicas: 1
  selector:
    matchLabels:
      app: cassandra-client
  template:
    metadata:
      labels:
        app: cassandra-client
    spec:
      containers:
      - image: bitnami/cassandra:4.0.1
        imagePullPolicy: IfNotPresent
        name: cassandra-client
`

const nginxDeployYaml = `
---
apiVersion: v1
kind: Secret
metadata:
  name: cassandra-secrets
type: Opaque
data:
  cassandra_password: Y2Fzc2FuZHJh
---
apiVersion: keda.sh/v1alpha1
kind: TriggerAuthentication
metadata:
  name: keda-trigger-auth-cassandra-secret
spec:
  secretTargetRef:
  - parameter: password
    name: cassandra-secrets
    key: cassandra_password
---
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app: nginx
  name: ${nginxDeploymentName}
spec:
  replicas: 0
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
      - image: nginx
        name: nginx
        ports:
        - containerPort: 80
---
apiVersion: keda.sh/v1alpha1
kind: ScaledObject
metadata:
  name: cassandra-scaledobject
spec:
  minReplicaCount: 0
  maxReplicaCount: 2
  pollingInterval: 1  # Optional. Default: 30 seconds
  cooldownPeriod: 1 # Optional. Default: 300 seconds
  scaleTargetRef:
    name: ${nginxDeploymentName}
  triggers:
  - type: cassandra
    metadata:
      username: "cassandra"
      clusterIPAddress: "cassandra.${cassandraNamespace}"
      consistency: "Quorum"
      protocolVersion: "4"
      port: "9042"
      keyspace: "${cassandraKeyspace}"
      query: "SELECT COUNT(*) FROM ${cassandraKeyspace}.${cassandraTableName};"
      targetQueryValue: "1"
      metricName: "${cassandraKeyspace}"
    authenticationRef:
      name: keda-trigger-auth-cassandra-secret
`

File: tests/scalers/cassandra/cassandra_test.go
//go:build e2e
// +build e2e

package cassandra_test

import (
	"bytes"
	"encoding/base64"
	"fmt"
	"testing"
	"text/template"

	"github.com/joho/godotenv"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
	"k8s.io/client-go/kubernetes"

	. "github.com/kedacore/keda/v2/tests/helper"
)

// Load environment variables from .env file
var _ = godotenv.Load("../../.env")

const (
	testName = "cassandra-test"
)

var (
	testNamespace      = fmt.Sprintf("%s-ns", testName)
	deploymentName     = fmt.Sprintf("%s-deployment", testName)
	scaledObjectName   = fmt.Sprintf("%s-so", testName)
	secretName         = fmt.Sprintf("%s-secret", testName)
	cassandraKeyspace  = "test_keyspace"
	cassandraTableName = "test_table"
	cassandraUsername  = "cassandra"
	cassandraPassword  = "cassandra"
	createKeyspace     = fmt.Sprintf("CREATE KEYSPACE IF NOT EXISTS %s WITH REPLICATION = {'class': 'NetworkTopologyStrategy', 'datacenter1' : '1'};", cassandraKeyspace)
	createTableCQL     = fmt.Sprintf("CREATE TABLE IF NOT EXISTS %s.%s (name text, surname text, age int, PRIMARY KEY (name, surname));", cassandraKeyspace, cassandraTableName)
	truncateData       = fmt.Sprintf("TRUNCATE %s.%s;", cassandraKeyspace, cassandraTableName)
	minReplicaCount    = 0
	maxReplicaCount    = 2
)

type templateData struct {
	TestNamespace           string
	DeploymentName          string
	ScaledObjectName        string
	SecretName              string
	Command                 string
	CassandraPasswordBase64 string
	CassandraKeyspace       string
	CassandraTableName      string
}

const (
	secretTemplate = `apiVersion: v1
kind: Secret
metadata:
  name: {{.SecretName}}
  namespace: {{.TestNamespace}}
data:
  cassandra_password: {{.CassandraPasswordBase64}}
`

	triggerAuthenticationTemplate = `apiVersion: keda.sh/v1alpha1
kind: TriggerAuthentication
metadata:
  name: keda-trigger-auth-cassandra-secret
  namespace: {{.TestNamespace}}
spec:
  secretTargetRef:
  - parameter: password
    name: {{.SecretName}}
    key: cassandra_password
`

	deploymentTemplate = `
apiVersion: apps/v1
kind: Deployment
metadata:
  name: {{.DeploymentName}}
  namespace: {{.TestNamespace}}
  labels:
    app: {{.DeploymentName}}
spec:
  replicas: 0
  selector:
    matchLabels:
      app: {{.DeploymentName}}
  template:
    metadata:
      labels:
        app: {{.DeploymentName}}
    spec:
      containers:
      - name: nginx
        image: ghcr.io/nginx/nginx-unprivileged:1.26
        ports:
        - containerPort: 80
`

	jobTemplate = `
apiVersion: batch/v1
kind: Job
metadata:
  name: client
  namespace: {{.TestNamespace}}
spec:
  template:
    spec:
      containers:
      - name: client
        image: "bitnami/cassandra"
        imagePullPolicy: Always
        command:
        - sh
        - -c
        - "{{.Command}}"
        env:
        - name: MAX_HEAP_SIZE
          value: 1024M
        - name: HEAP_NEWSIZE
          value: 100M
      restartPolicy: OnFailure
  backoffLimit: 4
`

	scaledObjectTemplate = `
apiVersion: keda.sh/v1alpha1
kind: ScaledObject
metadata:
  name: {{.ScaledObjectName}}
  namespace: {{.TestNamespace}}
  labels:
    app: {{.DeploymentName}}
spec:
  scaleTargetRef:
    name: {{.DeploymentName}}
  minReplicaCount: 0
  maxReplicaCount: 2
  pollingInterval: 1
  cooldownPeriod:  1
  triggers:
  - type: cassandra
    metadata:
      username: "cassandra"
      clusterIPAddress: "cassandra.{{.TestNamespace}}"
      consistency: "Quorum"
      protocolVersion: "4"
      port: "9042"
      keyspace: "{{.CassandraKeyspace}}"
      query: "SELECT COUNT(*) FROM {{.CassandraKeyspace}}.{{.CassandraTableName}};"
      targetQueryValue: "1"
      activationTargetQueryValue: "4"
      metricName: "{{.CassandraKeyspace}}"
    authenticationRef:
      name: keda-trigger-auth-cassandra-secret
`
	insertDataTemplateA = `BEGIN BATCH
    INSERT INTO {{.CassandraKeyspace}}.{{.CassandraTableName}} (name, surname, age) VALUES ('Mary', 'Paul', 30);
    INSERT INTO {{.CassandraKeyspace}}.{{.CassandraTableName}} (name, surname, age) VALUES ('James', 'Miller', 25);
    INSERT INTO {{.CassandraKeyspace}}.{{.CassandraTableName}} (name, surname, age) VALUES ('Lisa', 'Wilson', 29);
    APPLY BATCH;`

	insertDataTemplateB = `BEGIN BATCH
    INSERT INTO {{.CassandraKeyspace}}.{{.CassandraTableName}} (name, surname, age) VALUES ('Bob', 'Taylor', 33);
    INSERT INTO {{.CassandraKeyspace}}.{{.CassandraTableName}} (name, surname, age) VALUES ('Carol', 'Moore', 31);
    INSERT INTO {{.CassandraKeyspace}}.{{.CassandraTableName}} (name, surname, age) VALUES ('Richard', 'Brown', 23);
    APPLY BATCH;`
)

func TestCassandraScaler(t *testing.T) {
	// Create kubernetes resources
	kc := GetKubernetesClient(t)
	data, templates := getTemplateData()
	CreateNamespace(t, kc, testNamespace)
	// cleanup
	t.Cleanup(func() {
		uninstallCassandra(t)
		DeleteKubernetesResources(t, testNamespace, data, templates)
	})

	// setup cassandra
	installCassandra(t)
	setupCassandra(t, kc, data)

	// deploy test resources
	KubectlApplyMultipleWithTemplate(t, data, templates)

	assert.True(t, WaitForDeploymentReplicaReadyCount(t, kc, deploymentName, testNamespace, minReplicaCount, 60, 3),
		"replica count should be %d after 3 minutes", minReplicaCount)

	// test scaling
	testActivation(t, kc, data)
	testScaleOut(t, kc, data)
	testScaleIn(t, kc, data)
}

func installCassandra(t *testing.T) {
	_, err := ExecuteCommand("helm repo add bitnami https://charts.bitnami.com/bitnami")
	require.NoErrorf(t, err, "cannot execute command - %s", err)
	_, err = ExecuteCommand("helm repo update")
	require.NoErrorf(t, err, "cannot execute command - %s", err)
	_, err = ExecuteCommand(fmt.Sprintf("helm install cassandra --set resourcesPreset=none --set persistence.enabled=false --set dbUser.user=%s --set dbUser.password=%s --namespace %s bitnami/cassandra --wait", cassandraUsername, cassandraPassword, testNamespace))
	require.NoErrorf(t, err, "cannot execute command - %s", err)
}

func uninstallCassandra(t *testing.T) {
	_, err := ExecuteCommand(fmt.Sprintf("helm uninstall cassandra --namespace %s", testNamespace))
	assert.NoErrorf(t, err, "cannot execute command - %s", err)
}

func setupCassandra(t *testing.T, kc *kubernetes.Clientset, data templateData) {
	// Create the key space
	data.Command = fmt.Sprintf("cqlsh -u %s -p %s cassandra.%s --execute=\\\"%s\\\"", cassandraUsername, cassandraPassword, testNamespace, createKeyspace)
	KubectlApplyWithTemplate(t, data, "jobTemplate", jobTemplate)
	require.True(t, WaitForJobSuccess(t, kc, "client", testNamespace, 6, 10), "create database job failed")
	KubectlDeleteWithTemplate(t, data, "jobTemplate", jobTemplate)
	// Create the table
	data.Command = fmt.Sprintf("cqlsh -u %s -p %s cassandra.%s --execute=\\\"%s\\\"", cassandraUsername, cassandraPassword, testNamespace, createTableCQL)
	KubectlApplyWithTemplate(t, data, "jobTemplate", jobTemplate)
	require.True(t, WaitForJobSuccess(t, kc, "client", testNamespace, 6, 10), "create database job failed")
	KubectlDeleteWithTemplate(t, data, "jobTemplate", jobTemplate)

	t.Log("--- cassandra is ready ---")
}

func testActivation(t *testing.T, kc *kubernetes.Clientset, data templateData) {
	t.Log("--- testing activation ---")
	result, err := getCassandraInsertCmd(insertDataTemplateA)
	assert.NoErrorf(t, err, "cannot parse log - %s", err)
	data.Command = fmt.Sprintf("cqlsh -u %s -p %s cassandra.%s --execute=\\\"%s\\\"", cassandraUsername, cassandraPassword, testNamespace, result)
	KubectlApplyWithTemplate(t, data, "jobTemplate", jobTemplate)
	assert.True(t, WaitForJobSuccess(t, kc, "client", testNamespace, 6, 10), "insert job failed")
	KubectlDeleteWithTemplate(t, data, "jobTemplate", jobTemplate)

	AssertReplicaCountNotChangeDuringTimePeriod(t, kc, deploymentName, testNamespace, minReplicaCount, 60)
}

func testScaleOut(t *testing.T, kc *kubernetes.Clientset, data templateData) {
	t.Log("--- testing scale out ---")
	result, err := getCassandraInsertCmd(insertDataTemplateB)
	assert.NoErrorf(t, err, "cannot parse log - %s", err)
	data.Command = fmt.Sprintf("cqlsh -u %s -p %s cassandra.%s --execute=\\\"%s\\\"", cassandraUsername, cassandraPassword, testNamespace, result)
	KubectlApplyWithTemplate(t, data, "jobTemplate", jobTemplate)
	assert.True(t, WaitForJobSuccess(t, kc, "client", testNamespace, 6, 10), "insert job failed")
	KubectlDeleteWithTemplate(t, data, "jobTemplate", jobTemplate)

	assert.True(t, WaitForDeploymentReplicaReadyCount(t, kc, deploymentName, testNamespace, maxReplicaCount, 60, 3),
		"replica count should be %d after 3 minutes", maxReplicaCount)
}

func testScaleIn(t *testing.T, kc *kubernetes.Clientset, data templateData) {
	t.Log("--- testing scale in ---")

	data.Command = fmt.Sprintf("cqlsh -u %s -p %s cassandra.%s --execute=\\\"%s\\\"", cassandraUsername, cassandraPassword, testNamespace, truncateData)
	KubectlApplyWithTemplate(t, data, "jobTemplate", jobTemplate)
	assert.True(t, WaitForJobSuccess(t, kc, "client", testNamespace, 6, 10), "insert job failed")
	KubectlDeleteWithTemplate(t, data, "jobTemplate", jobTemplate)

	assert.True(t, WaitForDeploymentReplicaReadyCount(t, kc, deploymentName, testNamespace, minReplicaCount, 60, 3),
		"replica count should be %d after 3 minutes", minReplicaCount)
}

func getCassandraInsertCmd(insertDataTemplate string) (string, error) {
	tmpl, err := template.New("cassandra insert").Parse(insertDataTemplate)
	var tpl bytes.Buffer
	if err := tmpl.Execute(&tpl, templateData{CassandraKeyspace: cassandraKeyspace, CassandraTableName: cassandraTableName}); err != nil {
		return "", err
	}
	result := tpl.String()
	return result, err
}

func getTemplateData() (templateData, []Template) {
	return templateData{
			TestNamespace:           testNamespace,
			DeploymentName:          deploymentName,
			ScaledObjectName:        scaledObjectName,
			SecretName:              secretName,
			CassandraPasswordBase64: base64.StdEncoding.EncodeToString([]byte(cassandraPassword)),
			CassandraKeyspace:       cassandraKeyspace,
			CassandraTableName:      cassandraTableName,
		}, []Template{
			{Name: "secretTemplate", Config: secretTemplate},
			{Name: "triggerAuthenticationTemplate", Config: triggerAuthenticationTemplate},
			{Name: "deploymentTemplate", Config: deploymentTemplate},
			{Name: "scaledObjectTemplate", Config: scaledObjectTemplate},
		}
}
